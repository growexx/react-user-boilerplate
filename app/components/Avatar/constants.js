// component name as ga label
export const GA_LABEL_AVATAR = 'Avatar';
