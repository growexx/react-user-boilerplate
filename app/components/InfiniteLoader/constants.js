export const TABS = {
  TITLE: {
    TAB_ONE: 'List With Infinite Loader',
    TAB_TWO: 'List With Load More',
    TAB_THREE: 'List With Pagination',
  },
};
