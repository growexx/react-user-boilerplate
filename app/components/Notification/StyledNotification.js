import styled from 'styled-components';
import { Popover } from 'antd';
export const NotificationWrapper = styled.div`
  margin-right: 10px;
`;
export const StyledPopOver = styled(Popover)``;
