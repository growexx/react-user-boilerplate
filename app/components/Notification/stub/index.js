import React from 'react';
import {
  UserOutlined,
  CheckCircleOutlined,
  SmileOutlined,
  ExportOutlined,
  NumberOutlined,
} from '@ant-design/icons';
const NOTIFICATIONS = [
  {
    icon: <UserOutlined />,
    update: '70 new employees are shifted',
    timestamp: 1596119688264,
    read: false,
  },
  {
    icon: <CheckCircleOutlined />,
    update: 'Time to Take a Break, TADA!!!',
    timestamp: 1596119686811,
    read: false,
  },
  {
    icon: <SmileOutlined />,
    update: 'Time to Take a Break, TADA!!!',
    timestamp: 1596119686811,
    read: false,
  },
  {
    icon: <ExportOutlined />,
    update: 'Time to Take a Break, TADA!!!',
    timestamp: 1596119686811,
    read: false,
  },
  {
    icon: <NumberOutlined />,
    update: '70 new employees are shifted',
    timestamp: 1596119688264,
    read: false,
  },
];

export const TEST_IDS = {
  MARK_ALL_READ: 'Mark All As Read',
  BELL_ICON: 'BELL_ICON',
  EMPTY_CONTAINER: 'EMPTY_CONTAINER',
  INFINITE_SCROLLING: 'INFINITE_SCROLLING',
  NOTIFICATION_ITEM: 'NOTIFICATION_ITEM',
};

export const getNotificationsSuccessMock = () =>
  Promise.resolve({
    data: NOTIFICATIONS,
    status: 1,
  });
export const getNotificationsMockWithNoData = () =>
  Promise.resolve({
    data: [],
    status: 0,
  });
export const getNotificationsMockWithLessData = () =>
  Promise.resolve({
    data: [NOTIFICATIONS[0]],
    status: 1,
  });
export const getNotificationsFailureData = () =>
  Promise.reject(
    new Error({
      message: 'something went wrong',
    }),
  );

export const getNotificationsSuccessMockForClick = () =>
  Promise.resolve({
    data: [
      {
        icon:
          'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQZZrv3_PEnkdOIZvnr0COONt3kL7rSSq623dB3fyLCgT7GARpReF26nPOre6JCLHKu7KQ&usqp=CAU',
        update: '70 new employees are shifted',
        timestamp: 1596119688264,
        read: false,
        clickAction: 'http://localhost:3000/features',
      },
      {
        icon: <UserOutlined />,
        update: '70 new employees are shifted',
        timestamp: 1596119688264,
        read: true,
      },
      ...NOTIFICATIONS,
    ],
    status: 1,
  });
