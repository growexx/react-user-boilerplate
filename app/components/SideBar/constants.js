/*
 * Sidebar Constants
 * This file contains constants used in Sidebar component.
 */

import React from 'react';
import {
  UserOutlined,
  SmileOutlined,
  LoadingOutlined,
  ExportOutlined,
  NumberOutlined,
  FormOutlined,
  LockOutlined,
  PieChartOutlined,
  MessageOutlined,
  ShoppingOutlined,
  SearchOutlined,
} from '@ant-design/icons';
import {
  ROUTES,
  ROLE_BASED_SIDEBAR_MENU,
  ROLES,
  RESTRICTED_ROUTES,
} from 'containers/constants';

export const MenuItems = [
  {
    to: ROUTES.HOME,
    tabName: 'Home',
    icon: <SmileOutlined />,
  },
  {
    to: ROUTES.SAMPLE_FORM,
    tabName: 'Redux-Saga Form',
    icon: <FormOutlined />,
  },
  {
    to: ROUTES.LOADER,
    tabName: 'Loader Demo',
    icon: <LoadingOutlined />,
  },
  {
    to: ROUTES.EXPORT_DATA,
    tabName: 'Export Data To CSV',
    icon: <ExportOutlined />,
  },
  {
    to: ROUTES.NUMERAL_CONVERTER,
    tabName: 'Number Conversion Demo',
    icon: <NumberOutlined />,
  },
  {
    to: ROUTES.USERS,
    tabName: 'User Management',
    icon: <UserOutlined />,
  },
  {
    to: ROUTES.GITHUB_SEARCH,
    tabName: 'Github Search',
    icon: <SearchOutlined />,
  },
  {
    to: ROUTES.TEST_ADMIN_PAGE,
    tabName: 'Admin Page',
    icon: <LockOutlined />,
  },
  {
    to: ROUTES.CHARTS,
    tabName: 'Charts',
    icon: <PieChartOutlined />,
  },
  {
    to: ROUTES.MULTI_TAB_SUPPORT,
    tabName: 'Multi Tab Communication',
    icon: <MessageOutlined />,
  },
  {
    to: ROUTES.PRODUCTS,
    tabName: 'Products',
    icon: <ShoppingOutlined />,
  },
];

/**
 * Filters Sidebar menu based on role
 * @param {string} role
 */
export const GET_FILTERED_MENU_ITEM = role =>
  MenuItems.filter(item =>
    // Check if route is private then role has access to it
    RESTRICTED_ROUTES.includes(item.to)
      ? ROLE_BASED_SIDEBAR_MENU[role || ROLES.USER].includes(item.to)
      : true,
  );
