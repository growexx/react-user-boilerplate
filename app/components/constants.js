export const LAYOUT_CONFIG = {
  VERTICAL_OPTION_1: 1,
  VERTICAL_OPTION_2: 2,

  HORIZONTAL_OPTION_1: 3,
};

export const showLogoInSideBar = currentVariant =>
  ![
    LAYOUT_CONFIG.VERTICAL_OPTION_2,
    LAYOUT_CONFIG.HORIZONTAL_OPTION_1,
  ].includes(currentVariant);
