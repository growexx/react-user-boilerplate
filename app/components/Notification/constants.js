import { getNotificationsSuccessMock } from 'components/Notification/stub';
export const NOTIFICATION_LIMIT = 5;
export const getNotificationsMock = () => getNotificationsSuccessMock();
