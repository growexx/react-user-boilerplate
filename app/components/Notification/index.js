/**
 * Notification/index.js
 *
 * This is the Notification Component file.
 */
import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheck } from '@fortawesome/free-solid-svg-icons';
import { Waypoint } from 'react-waypoint';
import { Badge, List, Skeleton, Empty, notification, Button } from 'antd';
import { BellOutlined } from '@ant-design/icons';
import { TEST_IDS } from 'components/Notification/stub';
import {
  NotificationWrapper,
  StyledPopOver,
} from 'components/Notification/StyledNotification';
import {
  NOTIFICATION_LIMIT,
  getNotificationsMock,
} from 'components/Notification/constants';

class Notification extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      newItemsLoading: false,
      unreadCount: 0,
      notificationList: [],
      loading: false,
      hasMore: true,
    };
    this.newNotificationsCursor = 0;
  }

  loadNotifications = () => {
    const { notificationList, unreadCount } = this.state;
    setTimeout(() => {
      getNotificationsMock()
        .then(res => {
          if (res.status) {
            if (res.data.length !== NOTIFICATION_LIMIT) {
              this.setState({
                hasMore: false,
              });
            }
            this.setState({
              notificationList: [...notificationList, ...res.data],
              loading: false,
              unreadCount: res.data.length + unreadCount,
              newItemsLoading: false,
            });
          } else {
            this.setState({
              notificationList,
              unreadCount,
              loading: false,
              newItemsLoading: false,
            });
          }
        })
        .catch(err => {
          notification.error({ message: err.message });
          this.setState({
            notificationList,
            unreadCount,
            loading: false,
            newItemsLoading: false,
          });
        });
    }, 2000);
  };

  componentDidMount() {
    this.setState({
      loading: true,
    });
    this.loadNotifications();
  }

  getNewNotificationsLoader = loaderCount => {
    const { loading, newItemsLoading } = this.state;
    const loaderArray = [];
    // eslint-disable-next-line no-plusplus
    for (let index = 0; index < loaderCount; index++) {
      loaderArray.push(
        <Skeleton
          avatar
          title={false}
          loading={newItemsLoading || loading}
          active
          key={index}
        />,
      );
    }
    return <div className="newNotificationsLoader">{loaderArray}</div>;
  };

  handleReadCount = () => {
    const { notificationList } = this.state;
    const unreadCount = notificationList.filter(
      singleNotification => singleNotification.read === false,
    ).length;
    this.setState({
      unreadCount,
    });
  };

  handleNotificationClick = (item, index) => {
    const { notificationList } = this.state;
    const { read } = item;
    if (read === false) {
      const newItem = {
        ...item,
        read: true,
      };
      const currentItems = [...notificationList];
      currentItems[index] = newItem;
      this.setState(
        {
          notificationList: currentItems,
        },
        () => {
          this.handleReadCount();
        },
      );
    }
  };

  markAllNotificationsAsRead = () => {
    const { notificationList } = this.state;
    const updatedNotificationList = notificationList.map(
      updatedNotification => ({
        ...updatedNotification,
        read: true,
      }),
    );
    this.setState({
      notificationList: updatedNotificationList,
    });
  };

  handleMoreNotifications = () => {
    this.setState({
      newItemsLoading: true,
    });
    this.loadNotifications();
  };

  getNotificationContent = () => {
    const { notificationList, loading, hasMore } = this.state;
    const notificationsLength = notificationList.length;
    if (loading) {
      return this.getNewNotificationsLoader(10);
    }
    return (
      <>
        <List>
          {notificationsLength === 0 ? (
            <Empty data-testid={TEST_IDS.EMPTY_CONTAINER} />
          ) : (
            notificationList.map((item, index) => (
              // eslint-disable-next-line react/no-array-index-key
              <List.Item
                // eslint-disable-next-line react/no-array-index-key
                key={`${index}_${item}`}
                className={item.read === false ? 'readNotifications' : ''}
                onClick={() => this.handleNotificationClick(item, index)}
                data-testid={TEST_IDS.NOTIFICATION_ITEM}
              >
                <span className="notificationIcon">{item.icon}</span>
                <p className="notificationContent">{item.update}</p>
              </List.Item>
            ))
          )}
          {!loading && hasMore && (
            <Waypoint
              data-testid={TEST_IDS.INFINITE_SCROLLING}
              key={this.newNotificationsCursor}
              onEnter={this.handleMoreNotifications}
            >
              {this.getNewNotificationsLoader(2)}
            </Waypoint>
          )}
        </List>
      </>
    );
  };

  setMarkAllRead = () => {
    this.markAllNotificationsAsRead();
    this.setState({
      unreadCount: 0,
    });
  };

  getTitle = () => {
    const { unreadCount } = this.state;
    return (
      <>
        <p>Notifications</p>
        {unreadCount > 0 && (
          <>
            <FontAwesomeIcon
              icon={faCheck}
              title={TEST_IDS.MARK_ALL_READ}
              onClick={this.setMarkAllRead}
              data-testid={TEST_IDS.MARK_ALL_READ}
            />
          </>
        )}
      </>
    );
  };

  render() {
    const { unreadCount } = this.state;
    return (
      <NotificationWrapper>
        <div className="u-mr-1 u-d-inline-block">
          <Button
            type="text"
            data-testid="badge-Cart"
            className="btn-hover-none p-4"
          >
            <StyledPopOver
              placement="bottomLeft"
              content={this.getNotificationContent()}
              title={this.getTitle}
              overlayClassName="notificationPopoverContainer"
              trigger="click"
            >
              <Badge count={unreadCount} overflowCount={9} size="small">
                <BellOutlined
                  data-testid={TEST_IDS.BELL_ICON}
                  className="u-font-size-xlg"
                />
              </Badge>
            </StyledPopOver>
          </Button>
        </div>
      </NotificationWrapper>
    );
  }
}
export default Notification;
