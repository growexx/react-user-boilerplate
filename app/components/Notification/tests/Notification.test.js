import React from 'react';
import { fireEvent, render, wait } from 'react-testing-library';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import { ConnectedRouter } from 'connected-react-router/immutable';
import { createMemoryHistory } from 'history';
import configureStore from 'configureStore';
import 'jest-dom/extend-expect';
import {
  TEST_IDS,
  getNotificationsSuccessMock,
  getNotificationsSuccessMockForClick,
  getNotificationsFailureData,
  getNotificationsMockWithNoData,
  getNotificationsMockWithLessData,
} from 'components/Notification/stub';
import Notification from 'components/Notification/index';
import { getNotificationsMock } from 'components/Notification/constants';
jest.mock('components/Notification/constants');

describe('<Notification />', () => {
  const history = createMemoryHistory();
  const store = configureStore({}, history);
  it('should render notifications first time with success', async () => {
    getNotificationsMock.mockImplementation(() =>
      getNotificationsSuccessMock(),
    );
    const { getByTestId, getByText } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ConnectedRouter history={history}>
            <Notification />
          </ConnectedRouter>
        </IntlProvider>
      </Provider>,
    );
    fireEvent.click(getByTestId(TEST_IDS.BELL_ICON));
    await wait(() => {
      expect(getByText('Notifications')).toBeInTheDocument();
      expect(document.querySelector('.ant-skeleton')).toBeFalsy();
    });
    expect(getByTestId(TEST_IDS.MARK_ALL_READ)).toBeInTheDocument();
  });
  it('should render notifications first time with success and click on single notification', async () => {
    getNotificationsMock.mockImplementation(() =>
      getNotificationsSuccessMockForClick(),
    );
    const { getByTestId, getByText, getAllByTestId } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ConnectedRouter history={history}>
            <Notification />
          </ConnectedRouter>
        </IntlProvider>
      </Provider>,
    );
    fireEvent.click(getByTestId(TEST_IDS.BELL_ICON));
    await wait(() => {
      expect(getByText('Notifications')).toBeInTheDocument();
      expect(document.querySelector('.ant-skeleton')).toBeFalsy();
    });
    fireEvent.click(getAllByTestId(TEST_IDS.NOTIFICATION_ITEM)[0]);
    // reads one notification
    expect(getByText('4')).toBeInTheDocument();
    fireEvent.click(getAllByTestId(TEST_IDS.NOTIFICATION_ITEM)[1]);
    // reads one notification
    expect(getByText('3')).toBeInTheDocument();
  });
  it('should render notifications first time with success and click on mark all read', async () => {
    getNotificationsMock.mockImplementation(() =>
      getNotificationsSuccessMock(),
    );
    const { getByTestId, getByText, queryByTestId } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ConnectedRouter history={history}>
            <Notification />
          </ConnectedRouter>
        </IntlProvider>
      </Provider>,
    );
    fireEvent.click(getByTestId(TEST_IDS.BELL_ICON));
    await wait(() => {
      expect(getByText('Notifications')).toBeInTheDocument();
      expect(document.querySelector('.ant-skeleton')).toBeFalsy();
    });
    fireEvent.click(getByTestId(TEST_IDS.MARK_ALL_READ));
    expect(queryByTestId(TEST_IDS.MARK_ALL_READ)).not.toBeInTheDocument();
  });
  it('should render notifications first time with success and data less than limit', async () => {
    getNotificationsMock.mockImplementation(() =>
      getNotificationsMockWithLessData(),
    );
    const { getByTestId, getByText, queryByTestId } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ConnectedRouter history={history}>
            <Notification />
          </ConnectedRouter>
        </IntlProvider>
      </Provider>,
    );
    fireEvent.click(getByTestId(TEST_IDS.BELL_ICON));
    await wait(() => {
      expect(getByText('Notifications')).toBeInTheDocument();
      expect(document.querySelector('.ant-skeleton')).toBeFalsy();
    });
    expect(queryByTestId(TEST_IDS.INFINITE_SCROLLING)).not.toBeInTheDocument();
  });
  it('should render empty message', async () => {
    getNotificationsMock.mockImplementation(() =>
      getNotificationsMockWithNoData(),
    );
    const { getByTestId, getByText } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ConnectedRouter history={history}>
            <Notification />
          </ConnectedRouter>
        </IntlProvider>
      </Provider>,
    );
    fireEvent.click(getByTestId(TEST_IDS.BELL_ICON));
    await wait(() => {
      expect(getByText('Notifications')).toBeInTheDocument();
      expect(document.querySelector('.ant-skeleton')).toBeFalsy();
    });
    expect(getByTestId(TEST_IDS.EMPTY_CONTAINER)).toBeInTheDocument();
  });
  it('should catch the failure', async () => {
    getNotificationsMock.mockImplementation(() =>
      getNotificationsFailureData(),
    );
    const { getByTestId, getByText } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ConnectedRouter history={history}>
            <Notification />
          </ConnectedRouter>
        </IntlProvider>
      </Provider>,
    );
    fireEvent.click(getByTestId(TEST_IDS.BELL_ICON));
    await wait(() => {
      expect(getByText('Notifications')).toBeInTheDocument();
      expect(document.querySelector('.ant-skeleton')).toBeFalsy();
    });
    expect(getByTestId(TEST_IDS.EMPTY_CONTAINER)).toBeInTheDocument();
  });
  it('should render a div', () => {
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ConnectedRouter history={history}>
            <Notification />
          </ConnectedRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
