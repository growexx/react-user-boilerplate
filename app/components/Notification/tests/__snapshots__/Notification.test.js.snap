// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification /> should render a div 1`] = `
<div
  class="StyledNotification__NotificationWrapper-sc-103nxwt-0 dQHYeo"
>
  <div
    class="u-mr-3 u-d-inline-block"
  >
    <button
      class="ant-btn ant-btn-text"
      data-testid="badge-notification"
      type="button"
    >
      <span
        class="ant-badge"
      >
        <span
          aria-label="shopping-cart"
          class="anticon anticon-shopping-cart u-font-size-lg"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="shopping-cart"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M922.9 701.9H327.4l29.9-60.9 496.8-.9c16.8 0 31.2-12 34.2-28.6l68.8-385.1c1.8-10.1-.9-20.5-7.5-28.4a34.99 34.99 0 00-26.6-12.5l-632-2.1-5.4-25.4c-3.4-16.2-18-28-34.6-28H96.5a35.3 35.3 0 100 70.6h125.9L246 312.8l58.1 281.3-74.8 122.1a34.96 34.96 0 00-3 36.8c6 11.9 18.1 19.4 31.5 19.4h62.8a102.43 102.43 0 00-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7h161.1a102.43 102.43 0 00-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7H923c19.4 0 35.3-15.8 35.3-35.3a35.42 35.42 0 00-35.4-35.2zM305.7 253l575.8 1.9-56.4 315.8-452.3.8L305.7 253zm96.9 612.7c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 01-31.6 31.6zm325.1 0c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 01-31.6 31.6z"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
  <span
    aria-label="bell"
    class="anticon anticon-bell"
    role="img"
  >
    <svg
      aria-hidden="true"
      data-icon="bell"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M816 768h-24V428c0-141.1-104.3-257.7-240-277.1V112c0-22.1-17.9-40-40-40s-40 17.9-40 40v38.9c-135.7 19.4-240 136-240 277.1v340h-24c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h216c0 61.8 50.2 112 112 112s112-50.2 112-112h216c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM512 888c-26.5 0-48-21.5-48-48h96c0 26.5-21.5 48-48 48zM304 768V428c0-55.6 21.6-107.8 60.9-147.1S456.4 220 512 220c55.6 0 107.8 21.6 147.1 60.9S720 372.4 720 428v340H304z"
      />
    </svg>
  </span>
  <div
    data-testid="badge-cart-drawer"
    data-visible="false"
  />
</div>
`;
